KAFKA-7639: Read one request at a time from socket to avoid OOM

rajinisivaram opened the pull request:
Broker's Selector currently reads all requests available on the socket when the socket is ready for read. These are queued up as staged receives. This can result in excessive memory usage when socket read buffer size is large. We mute the channel and stop reading any more data until all the staged requests are processed. This behaviour is slightly inconsistent since for the initial read we drain the socket buffer, allowing it to get filled up again, but if data arrives slighly after the initial read, then we dont read from the socket buffer until pending requests are processed. 

To avoid holding onto requests for longer than required, this PR removes staged receives and reads one request at a time even if more data is available in the socket buffer. This is especially useful for produce requests which may be large and may take long to process. Additional data read from the socket for SSL is limited to the pre-allocated intermediate SSL buffers.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rajinisivaram committed:
KAFKA-7639: Read one request at a time from socket to avoid OOM

rajinisivaram commented:
@junrao This is an initial PR to read one request at-a-time, so we only have one entry in staged receives queue at any time from each connection on the broker (apart from re-authentication, which I will address separately later if possible). I have checked that existing tests work, but haven't yet written new tests for this. I first wanted to check if you think there may be any issues with this approach.

1) For PLAINTEXT, we read exact bytes into NetworkReceive, so we can stop as soon as channel.read() returns a completed receive. The socket buffer could have more data to read, which we will read after sending the response of the current receive. 
2) For SSL, we can't read exact number of encrypted bytes from the network. But we allocate exact buffers for unencrypted data. We read data from the socket into `SslTransportLayer#netReadBuffer`, decrypt into  `SslTransportLayer#appReadBuffer` and copy exact bytes from `appReadBuffer` to the current `NetworkReceive` (4 bytes followed by `size` bytes). So even though we may have read part of the next request from the socket into `netReadBuffer`, we dont need to process that data once we have a completed receive. So similar to PLAINTEXT, I don't think we need to store multiple staged receives in Selector for each connection.

With this PR, we have only one outstanding `NetworkReceive` at any one time from each connection. For SSL, we may also have additional data in the `SslTransportLayer` buffers, but we will process these after the current request is processed.

Have I missed something?

junrao commented:
@rajinisivaram : Thanks for the patch. The tricky thing is for #2. Let's say SslTransportLayer received 2 complete requests in appReadBuffer in a single read call and we only take one out to the channel in the Selector.poll() call. If there are no new requests for the socket, the selector won't select this key (channel) for read since there are no more bytes from the socket (even though there are bytes in the SslTransportLayer's appReadBuffer). So, the 2 issues are (1) the reading of a request may be delayed up to the selector timeout, (2) even when select.poll() returns, we don't know which key has data in appReadBuffer.

rajinisivaram commented:
@junrao Thanks for the explanation - now I remember why we had staged receives in the first place. But I think we don't need it any more because when we added memory pools, we also added `keysWithBufferedRead`. This tracks channels which have any data in SslTransportLayer `netReadBuffer` or `appReadBuffer`. And these are handled similar to staged receives (without delaying for selector timeout and by processing independently of whether there is new data to read). Actually I think we are polling with timeout=0 more often than we should for SSL, but I will fix that too as part of this PR.

So I think we can remove `stagedReceives` and read at most one request at a time for both PLAINTEXT and SSL when channel is not muted (apart from the data in SSL buffers). And `keysWithBufferedRead` will handle remaining data in buffers when channel is unmuted after previous request is processed. What do you think?

junrao commented:
@rajinisivaram : Yes, I agree that keysWithBufferedRead can potentially be used to replace stagedReceives.

ijuma commented:
Nice find @rajinisivaram! Removing `stagedReceives` would be a nice simplification.

ijuma commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -106,7 +105,7 @@
     private final List<Send> completedSends;
     private final List<NetworkReceive> completedReceives;
-    private final Map<KafkaChannel, Deque<NetworkReceive>> stagedReceives;
+    private final Set<KafkaChannel> channelsWithCompletedReceives;
```
@rajinisivaram did we consider changing `completedReceives` to be a `LinkedHashMap` from `channel` to `NetworkReceive`?

rajinisivaram replied:
@ijuma Thanks for the review, updated.

rajinisivaram commented:
@ijuma I was trying not to change the `Selectable` interface, but I suppose it is ok since it is not public. Will update.

ijuma commented:
@junrao would you have time to review this PR? It seems like a nice win.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -558,12 +554,21 @@ void pollSelectionKeys(Set<SelectionKey> selectionKeys,
-                attemptRead(key, channel);
+                //if channel is ready and has bytes to read from socket or buffer, and has no
+                //previous completed receive then read from it
+                if (channel.ready() && (key.isReadable() || channel.hasBytesBuffered()) && !hasCompletedReceive(channel)
```
Do we need to check hasCompletedReceive here? completedReceives is cleared after each poll and the same socket key is never polled more than once within each poll.

rajinisivaram replied:
We need the check because the reauthentication code above may add a completed receive.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -457,6 +453,10 @@ public void poll(long timeout) throws IOException {
 
+        // Update expiry time for closing channels whose buffered receive have been processed
+        if (idleExpiryManager != null)
+          completedReceives.keySet().forEach(channel -> idleExpiryManager.update(channel.id(), endSelect));
```
Hmm, not sure that I understand the code here. We call clear() earlier, which clears completedReceives. So this seems to be always a no-op?

rajinisivaram replied:
We needed something like this for staged receives, but we don't require this now, removed.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -558,12 +554,21 @@ void pollSelectionKeys(Set<SelectionKey> selectionKeys,
+                }
+                if (channel.ready() && channel.state() == ChannelState.NOT_CONNECTED)
+                    channel.state(ChannelState.READY);
+                if (!hasCompletedReceive(channel)) {
```
Is the hasCompletedReceive check needed? completedReceives is cleared after each poll and the same socket key is never polled more than once within each poll.

rajinisivaram replied:
Not required, removed.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -558,12 +554,21 @@ void pollSelectionKeys(Set<SelectionKey> selectionKeys,
-                    responsesReceivedDuringReauthentication.forEach(receive -> addToStagedReceives(channel, receive));
+                }
+                if (channel.ready() && channel.state() == ChannelState.NOT_CONNECTED)
+                    channel.state(ChannelState.READY);
```
Not sure why we need to set the channel state here. If channel.ready() is true, it seems that channel.state() should be READY?

rajinisivaram replied:
We don't currently set channel state to READY (and nothing relies on it). But since it is good to set the state anyway, I added this code.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java:
```diff
@@ -598,12 +598,10 @@ public ApiVersionsResponse apiVersionsResponse() {
          *         re-authentication, if any
          */
-        public List<NetworkReceive> getAndClearResponsesReceivedDuringReauthentication() {
+        public Optional<NetworkReceive> pollResponseReceivedDuringReauthentication() {
```
Should we change the comment above on responses?

junrao commented on clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java:
```diff
@@ -620,8 +616,8 @@ public boolean serverAuthenticationSessionExpired(long nowNanos) {
      */
-    public List<NetworkReceive> getAndClearResponsesReceivedDuringReauthentication() {
-        return authenticator.getAndClearResponsesReceivedDuringReauthentication();
+    public Optional<NetworkReceive> pollResponseReceivedDuringReauthentication() {
```
Should we change the comment above on responses?

junrao commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -105,8 +104,7 @@
     private final List<Send> completedSends;
-    private final List<NetworkReceive> completedReceives;
-    private final Map<KafkaChannel, Deque<NetworkReceive>> stagedReceives;
+    private final LinkedHashMap<KafkaChannel, NetworkReceive> completedReceives;
```
Is the ordering of the map important?

rajinisivaram replied:
No, but we used to return a List before, so this retains the order.

junrao commented on clients/src/test/java/org/apache/kafka/common/network/NioEchoServer.java:
```diff
@@ -47,6 +48,7 @@
 import java.util.Map;
 import java.util.Set;
 
+
```
Is this needed?

rajinisivaram replied:
no, removed

junrao commented on clients/src/test/java/org/apache/kafka/common/network/SelectorTest.java:
```diff
@@ -464,69 +447,126 @@ private void verifyImmediatelyConnectedException(Selector selector, String id) t
+     * Verifies that sockets with incoming data available are not expired.
+     * For PLAINTEXT, pending receives are always read from socket without any buffering, so this
+     * test is only verifying that channels are not expired while there is data to read from socket.
+     * For SSL, pending receives may also be in SSL netReadBuffer or appReadBuffer. So the test verifies
```
Are we testing the SSL in the test? It seems that the security protocol is always plaintext.

rajinisivaram replied:
We have `SslSelectorTest` that extends this class, so all the tests are run with PLAINTEXT and SSL.

junrao commented on clients/src/test/java/org/apache/kafka/common/network/SelectorTest.java:
```diff
@@ -464,69 +447,126 @@ private void verifyImmediatelyConnectedException(Selector selector, String id) t
+            assertEquals(channel.id(), selector.disconnected().keySet().iterator().next());
+            maxReceiveCountAfterClose = Math.max(maxReceiveCountAfterClose, receiveCount);
+        }
+        assertTrue("Too few receives after close: " + maxReceiveCountAfterClose, maxReceiveCountAfterClose >= 5);
```
I guess the reason that we don't know the exact ReceiveCountAfterClose is that we don't know how much data is in the socket buffer when we close the connections on the server. But how do we guarantee that we always have at least 5 receives?

rajinisivaram replied:
The loop above retries until we have at least 5 but limits the number of retries to 100, so this is checking that it worked within that retry limit.

junrao commented on clients/src/test/java/org/apache/kafka/common/network/SelectorTest.java:
```diff
@@ -464,69 +447,126 @@ private void verifyImmediatelyConnectedException(Selector selector, String id) t
+        TestUtils.waitForCondition(() -> {
+            try {
+                selector.poll(100);
+                return !selector.disconnected().isEmpty() || !selector.completedReceives().isEmpty();
```
Is the second test needed since there will never be a complete receive?

rajinisivaram replied:
Removed.

junrao commented on core/src/test/scala/unit/kafka/network/SocketServerTest.scala:
```diff
@@ -257,37 +272,41 @@ class SocketServerTest extends JUnitSuite {
+      processRequest(overrideServer.requestChannel, request0)
+      assertTrue("Channel not open", openChannel(request0, overrideServer).nonEmpty)
+      assertEquals(openChannel(request0, overrideServer), openOrClosingChannel(request0, overrideServer))
+      TestUtils.waitUntilTrue(() => !openChannel(request0, overrideServer).head.isMute, "Failed to unmute channel")
```
openChannel(request0, overrideServer).head =>openChannel(request0, overrideServer).get ?

junrao commented on core/src/test/scala/unit/kafka/network/SocketServerTest.scala:
```diff
@@ -1301,5 +1565,74 @@ class SocketServerTest extends JUnitSuite {
+    }
+  }
+
+  private class ProxyServer(socketServer: SocketServer) {
```
Could we add a bit comment on what ProxyServer does?

junrao commented on core/src/test/scala/unit/kafka/network/SocketServerTest.scala:
```diff
@@ -916,29 +979,206 @@ class SocketServerTest extends JUnitSuite {
+    verifyRemoteCloseWithBufferedReceives(numComplete = 3, hasIncomplete = false, responseRequiredIndex = 1, makeClosing = false)
+  }
+
+  private def verifyRemoteCloseWithBufferedReceives(numComplete: Int,
```
It's kind of hard to understand the logic in this method. Could we add a bit comment on what this method does and also add some additional comments on each of the major steps?

junrao reviewed (commented):
@rajinisivaram : Thanks for the patch. Left a few comments below.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@junrao @ijuma Thanks for the reviews. Sorry that was almost an year ago! I have rebased and addressed the comments.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -429,10 +426,9 @@ public void send(Send send) {
-     * reading a channel we read as many responses as we can and store them into "stagedReceives" and pop one response during
-     * the poll to add the completedReceives. If there are any active channels in the "stagedReceives" we set "timeout" to 0
-     * and pop response and add to the completedReceives.
+     * in the buffer. To overcome this issue we added "keysWithBufferedRead" map which tracks channels which have data in the SSL
```
This is an existing issue. The line above, "we've have" => "we have".

junrao commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -454,7 +450,7 @@ public void poll(long timeout) throws IOException {
         boolean dataInBuffers = !keysWithBufferedRead.isEmpty();
 
-        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty() || (madeReadProgressLastCall && dataInBuffers))
+        if (!immediatelyConnectedKeys.isEmpty() || (madeReadProgressLastCall && dataInBuffers))
```
This is probably not a concern, but the current logic means that the pending reads in closingChannels could be delayed by the selector timeout.

rajinisivaram replied:
Yes, I can't think of an easy way to fix that unfortunately.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java:
```diff
@@ -611,12 +611,10 @@ public ApiVersionsResponse apiVersionsResponse() {
          *         re-authentication, if any
          */
-        public List<NetworkReceive> getAndClearResponsesReceivedDuringReauthentication() {
+        public Optional<NetworkReceive> pollResponseReceivedDuringReauthentication() {
```
The comments above still mention responses. They should be response?

junrao commented on clients/src/test/java/org/apache/kafka/common/network/SelectorTest.java:
```diff
@@ -522,71 +506,126 @@ private void verifyImmediatelyConnectedException(Selector selector, String id) t
         while (selector.disconnected().isEmpty()) {
             time.sleep(6000); // The max idle time is 5000ms
-            selector.poll(0);
+            selector.poll(completedReceives == expectedReceives ? 0 : 1000);
```
Hmm, where is the logic to close the connection?

rajinisivaram replied:
We update the mock time above and expect the channel to be closed due to idle timeout.

junrao commented on core/src/test/scala/unit/kafka/network/SocketServerTest.scala:
```diff
@@ -519,23 +550,91 @@ class SocketServerTest {
+
+      socket.close()
+      proxyServer.serverConnSocket.close()
+      TestUtils.waitUntilTrue(() => proxyServer.clientConnSocket.isClosed, "Client socket not closed")
```
The default wait time in waitUntilTrue() is 15 secs, which is probably too long.

junrao commented on core/src/test/scala/unit/kafka/network/SocketServerTest.scala:
typo remvoved

junrao reviewed (approved):
@rajinisivaram : Thanks for the updated patch. The logic in Selector does look simpler. LGTM. Just a few minor comments below.

rajinisivaram committed:
Address review comments

rajinisivaram committed:
Rebase and remove references to CoreUtils.runnable

rajinisivaram commented:
@junrao Thanks for the review. I have addressed the comments.

rajinisivaram commented:
Test failure `kafka.admin.ReassignPartitionsClusterTest.shouldListMovingPartitionsThroughApi` not related. Merging to trunk.
